src/protocol.h

```diff
@@ -71,6 +71,8 @@
 #define PROTMESSID_CLM_REQ_VERSION_AND_OS     1012 // request version number and operating system
 #define PROTMESSID_CLM_CONN_CLIENTS_LIST      1013 // channel infos for connected clients
 #define PROTMESSID_CLM_REQ_CONN_CLIENTS_LIST  1014 // request the connected clients list
+#define PROTMESSID_CLM_REQ_CHANNEL_LEVEL_LIST 1015 // request the channel level list
```

Comment:
Hm, I can see that you have used a connection-less message also for the enabling of the feature. I have not studied your code in great detail yet, but if this is the case, we can have issues with that. E.g. if you have temporary network issues and you switch on the LEDs at the client at that moment, it may happen that the REQ_CHANNEL_LEVEL does not make it through to the server. In that case the GUI control shows that LED is enabled but the server does not send any levels.
So you should use a normal message for enabling the feature (for normal messages we have a check for lost packets and a re-send mechanism) and, what you already did, a connection-less message for the transmit of the levels from the server to the client to get the lowest network traffic for the updates. If one connection-less packet with the levels does not make it to the client is not an issue at all since we have continuous updates of the levels.
Do you agree with this?

Reply:
Oh right, yes, that makes sense.

Reply:
Hm.  OK, looking through the code, I don't spot an example I can directly follow for the request.

For the CLM, the client "implicitly" sends its address, which the server uses to determine the channel in question.

For the standard request, clients do not appear to send their address -- but I don't see how the server then knows _which_ channel the request is for - some how, the "emit" just seems to get there.

Can you explain what happens - i.e. how I get the client opt in to the server right channel?

Reply:
You could take a look at, e.g., the "CreateJitBufMes". No need to set the addess since the client is in connected state and the address is already stored in the channel.

Reply:
The problem is that you have to find the correct place to send the message. You sould put it in "OnNewConnection()" to update the current setting of the client in the server.

Reply:
Yes, it's there.  src/client.cpp:
```
 288     CreateServerJitterBufferMessage();
 289
 290     // send opt-in / out for Channel Level updates
 291     SetDisplayChannelLevels ( bDisplayChannelLevels );
```

Reply:
So...
CClient::CreateServerJitterBufferMessage()
-> Channel::CreateJitBufMes( ... ) -> Protocol::CreateJitBufMes( ... )
-> Protocol::CreateAndSendMessage (  )
...
-> CProtocol::EvaluateJitBufMes ( ... ) -> emit ChangeJittBufSize ( ... )
-> CProtocol SIGNAL ( ChangeJittBufSize ( ... ) ) -> CChannel SLOT ( OnJittBufSizeChange )

But which channel?  Magic happened.  server.cpp never sees the signal.

CChannel::OnJittBufSizeChange then leads through into
```
src/server.h:574:    void OnServerAutoSockBufSizeChangeCh0  ( int iNNumFra ) { vecChannels[0].CreateJitBufMes  ( iNNumFra ); }
src/server.h:575:    void OnServerAutoSockBufSizeChangeCh1  ( int iNNumFra ) { vecChannels[1].CreateJitBufMes  ( iNNumFra ); }
src/server.h:576:    void OnServerAutoSockBufSizeChangeCh2  ( int iNNumFra ) { vecChannels[2].CreateJitBufMes  ( iNNumFra ); }
s
```
But, again, I still don't see how the server knew which channel received the message to route the signal correctly.

Reply:
> CChannel SLOT ( OnJittBufSizeChange ) [...] But which channel?

If that message would be your new message. Then in function OnJittBufSizeChange() in the "if ( bIsServer )" section you would update your new variable bChannelLevelsRequired.
But maybe I did not fully understand your question.

Reply:
BTW: The "AutoSockBufSize" is something different and should not be used as a reference.

Reply:
So it sounds like there is magic happening in the "connected" message handling that "just knows" where the message came from so it goes to the right channel?  OK, if it just works, it just works :).

Reply:
Not really magic :-). The server has a channel for each connected client. The client just only "talks" to its own channel at the server.

Reply:
Yeah, just took some mental context switching for "connected" channel over UDP to make sense...  It's "magic" in that it's hidden in the protocol layer implementation and the upper layer (message handling) doesn't need to worry.